src/execution/index/art/iterator.cpp

```diff
@@ -0,0 +1,281 @@
+		if (cur_node->type == NodeType::NLeaf) {
+			// Pop Leaf (We must pop the prefix size + the key to the node (unless we are popping the root)
+			idx_t elements_to_pop = cur_node->prefix.Size() + (nodes.size() != 1);
+			cur_key.Pop(elements_to_pop);
+			nodes.pop();
+		}
+	}
+
+	// Look for the next leaf
+	while (!nodes.empty()) {
+		// cur_node
+		auto &top = nodes.top();
+		Node *node = top.node;
+		if (node->type == NodeType::NLeaf) {
+			// found a leaf: move to next node
+			last_leaf = (Leaf *)node;
+			return true;
+		}
+		// Find next node
+		top.pos = node->GetNextPos(top.pos);
+		if (top.pos != DConstants::INVALID_INDEX) {
+			// add key-byte of the new node
+			PushKey(node, top.pos);
+			auto next_node = node->GetChild(*art, top.pos);
+			// add prefix of new node
+			for (idx_t i = 0; i < next_node->prefix.Size(); i++) {
+				cur_key.Push(next_node->prefix[i]);
+			}
+			// next node found: push it
+			nodes.push(IteratorEntry(next_node, DConstants::INVALID_INDEX));
+		} else {
+			// no node found: move up the tree and Pop prefix and key of current node
+			auto cur_node = nodes.top().node;
+			idx_t elements_to_pop = cur_node->prefix.Size() + (nodes.size() != 1);
+			cur_key.Pop(elements_to_pop);
+			nodes.pop();
+		}
+	}
+	return false;
+}
+
+bool Iterator::Bound(Node *node, Key &key, bool inclusive) {
+	bool equal = false;
+	if (!node) {
+		return false;
+	}
+	idx_t depth = 0;
+	while (true) {
+		nodes.push(IteratorEntry(node, 0));
+		auto &top = nodes.top();
+		// reconstruct the prefix
+		for (idx_t i = 0; i < top.node->prefix.Size(); i++) {
+			cur_key.Push(top.node->prefix[i]);
+		}
+		if (!equal) {
+			while (node->type != NodeType::NLeaf) {
+				auto min_pos = node->GetMin();
+				PushKey(node, min_pos);
+				nodes.push(IteratorEntry(node, min_pos));
+				node = node->GetChild(*art, min_pos);
+				// reconstruct the prefix
+				for (idx_t i = 0; i < node->prefix.Size(); i++) {
+					cur_key.Push(node->prefix[i]);
+				}
+				auto &c_top = nodes.top();
+				c_top.node = node;
+			}
+		}
+		if (node->type == NodeType::NLeaf) {
+			// found a leaf node: check if it is bigger or equal than the current key
+			auto leaf = static_cast<Leaf *>(node);
+			last_leaf = leaf;
+			// if the search is not inclusive the leaf node could still be equal to the current value
+			// check if leaf is equal to the current key
+			if (cur_key == key) {
+				// if it's not inclusive check if there is a next leaf
+				if (!inclusive && !Next()) {
+					return false;
+				} else {
+					return true;
+				}
+			}
+
+			if (cur_key > key) {
+				return true;
+			}
+			// Leaf is lower than key
+			// Check if next leaf is still lower than key
+			while (Next()) {
+				if (cur_key == key) {
+					// if it's not inclusive check if there is a next leaf
+					if (!inclusive && !Next()) {
+						return false;
+					} else {
+						return true;
+					}
+				} else if (cur_key > key) {
+					// if it's not inclusive check if there is a next leaf
```

Comment:
What do you mean here? that the key comparison is copied?
I think it's annoying to put it in a method since its returns are part of the logic of the function